dev/tools/lib/roll_dev.dart

```diff
@@ -155,29 +118,79 @@ void main(List<String> args) {
       'to the "dev" channel.');
     stdout.write('Are you? [yes/no] ');
     if (stdin.readLineSync() != 'yes') {
-      runGit('tag -d $version', 'remove the tag you did not want to publish');
+      git.run('tag -d $version', 'remove the tag you did not want to publish');
       print('The dev roll has been aborted.');
-      exit(0);
+      return false;
     }
   }
 
-  runGit('push $origin $version', 'publish the version');
-  runGit('push $origin HEAD:dev', 'land the new version on the "dev" branch');
+  git.run('push $origin $version', 'publish the version');
+  git.run(
+    'push ${force ? "--force " : ""}$origin HEAD:dev',
+    'land the new version on the "dev" branch',
+  );
   print('Flutter version $version has been rolled to the "dev" channel!');
+  return true;
 }
 
-String getFullTag() {
+ArgResults parseArguments(ArgParser argParser, List<String> args) {
+  argParser.addOption(
+    kIncrement,
+    help: 'Specifies which part of the x.y.z version number to increment. Required.',
+    valueHelp: 'level',
+    allowed: <String>[kX, kY, kZ],
+    allowedHelp: <String, String>{
+      kX: 'Indicates a major development, e.g. typically changed after a big press event.',
+      kY: 'Indicates a minor development, e.g. typically changed after a beta release.',
+      kZ: 'Indicates the least notable level of change. You normally want this.',
+    },
+  );
+  argParser.addOption(
+    kCommit,
+    help: 'Specifies which git commit to roll to the dev branch. Required.',
+    valueHelp: 'hash',
+    defaultsTo: null, // This option is required
+  );
+  argParser.addOption(
+    kOrigin,
+    help: 'Specifies the name of the upstream repository',
+    valueHelp: 'repository',
+    defaultsTo: 'upstream',
+  );
+  argParser.addFlag(
+    kForce,
+    abbr: 'f',
+    help: 'Force push. Necessary when the previous release had cherry-picks.',
+    negatable: false,
+  );
+  argParser.addFlag(
+    kJustPrint,
+    negatable: false,
+    help:
+        "Don't actually roll the dev channel; "
+        'just print the would-be version and quit.',
+  );
+  argParser.addFlag(kYes, negatable: false, abbr: 'y', help: 'Skip the confirmation prompt.');
+  argParser.addFlag(kHelp, negatable: false, help: 'Show this help message.', hide: true);
+
+  return argParser.parse(args);
+}
+
+/// Obtain the version tag of the previous dev release.
+String getFullTag(Git git) {
   const String glob = '*.*.*-*.*.pre';
-  return getGitOutput(
-    'describe --match $glob --first-parent --long --tags',
+  // describe the latest dev release
+  const String ref = 'refs/heads/dev';
+  return git.getOutput(
+    'describe --match $glob --exact-match --tags $ref',
     'obtain last released version number',
   );
 }
 
 Match parseFullTag(String version) {
   // of the form: x.y.z-m.n.pre-c-g<revision>
   final RegExp versionPattern = RegExp(
-    r'^(\d+)\.(\d+)\.(\d+)-(\d+)\.(\d+)\.pre-(\d+)-g([a-f0-9]+)$');
+    r'^(\d+)\.(\d+)\.(\d+)-(\d+)\.(\d+)\.pre$');
```

Comment:
This is part of change 2. Instead of describing HEAD, we grab the exact tag at the tip of `refs/heads/dev`.